### ISSUE TITLE
Feature request: dual numbers as dtype

### ISSUE THREAD

**Issue by tBuLi** [numpy/numpy#16967]
I think it would be great to add [dual number](https://en.wikipedia.org/wiki/Dual_number) support to numpy. Dual numbers are numbers a + b ε where ε^2 = 0. These numbers have great utility in automatic differentiation, since from a Taylor series it directly follows that

f(x+ε) = f(x) + εf'(x)

Therefore one only has to look at the dual component of f(x + ε) to obtain the first derivative. Although there are already some libraries for this in python, it would obviously be much faster if this was in numpy itself, and it would also greatly increase the number of applications. The amount of new code would probably be similar or smaller to that of adding complex numbers themselves.

For automatic derivation of complex functions the coefficients of the dual number should be complex. And higher order derivatives can be evaluated using nested dual numbers, though that is probably less important for most users so if that is to exotic for a new dtype then that is not essential.

### Example code:

Derivative of f(x) = x^2:
```python
>>> import numpy as np
>>> x = np.array([np.dual(3,1), np.dual(5,1)])
>>> x**2
np.array([np.dual(9,6), np.dual(25,10)])
```
Derivative of f(z) = z^2, z complex:
```python
>>> y = np.dual(3 + 4j, 1)
>>> y**2
np.dual(-7 + 24j, 6 + 8j)
```
Second derivative of f(x) = x^2 using nested dual (might be to exotic for a dtype?):
```python
>>> x = np.dual(np.dual(3, 1), np.dual(1, 0))
>>> x**2
np.dual(np.dual(9, 6), np.dual(6, 2))
```
I'm curious if there is interest in adding this new feature, or if it was decided in the past not to do this, why that is? I've searched but couldn't find any history on dual numbers in numpy.

**Comment by charris** [comment]
I've toyed with the idea, I didn't know the idea had a name, thanks for that :) This sounds like something that would be useful and could maybe be added when the dtype work finishes, but I'm not convinced NumPy is the right place for it. We have done a fair amount of work that allows numpy functions to be overridden (__array_ufunc__) and a module making use of that might be a better way to go.

**Comment by tBuLi** [comment]
Glad I could help ;). I've looked up `__array_ufunc__` and that does seem like it could also work. The only thing I'm afraid of there is that (if I've understood correctly) the behavior would have to be implemented for every function with which the user could try and call the subclassed array. 

But taking the bottom up approach and looking at the C code I have the feeling that once a couple of functions like
```C
extern double c_abs(complex *);
extern void c_cos(complex *, complex *);
extern void c_div(complex *, complex *, complex *);
extern void c_exp(complex *, complex *);
extern void c_log(complex *, complex *);
extern void c_sin(complex *, complex *);
extern void c_sqrt(complex *, complex *);
extern void pow_ci(complex *, complex *, integer *);
extern void r_cnjg(complex *, complex *);
extern float r_imag(complex *);
```
have been defined for dual numbers then all other code will automatically follow. So it would seem to me that the bottom up approach is ultimately much more powerful since it will work in functions we have never anticipated.

And to play devils advocate: why should only complex numbers be recognized as fundamental number systems whereas dual numbers and split-complex numbers are discriminated against? It is true that these numbers are not always divisible, but matrices are also not always invertable but they are still pretty useful :). So I advocate for equal rights for all numbers of the form a + b i, whether i^2 = -1, i^2 = 0 or i^2 = 1.

**Comment by eric-wieser** [comment]
> And to play devils advocate: why should only complex numbers be recognized as fundamental number systems whereas dual numbers and split-complex numbers are discriminated against

I'm afraid this question has an easy answer: `complex` numbers are native to both python and C, unlike the others.

**Comment by mhvk** [comment]
@tBuLi - from my experience with adding `Quantity` to astropy, having numerical functions defined goes most of the way towards working with many parts of numpy. I don't think adding `__array_ufunc__` is that much more work than defining new C code...

**Comment by tBuLi** [comment]
@eric-wieser: Fair enough, I guess I should just keep taking this message up the food chain ;).

@mhvk Thank you for sharing that experience, that is comforting to hear. I will try implementing it that way then. Perhaps that will already be sufficient for my needs.

But having this number system accepted on the deeper levels will remain my dream :).

**Comment by ev-br** [comment]
I suggested a student to look at these based off the quaternion example some time around end of May. So far I did not hear from him so not sure if it's progessing :-).

**Comment by MichaelTiemannOSC** [comment]
I just talked about this with folks in the Newcomers meeting. Complex numbers are implemented in low-level code for a reason, but there's no reason we cannot simply create an ArrayExtension for Dual Numbers and let performance freaks do what they want to speed things up.

**Comment by seberg** [comment]
You can do an array-like for it with protocols, it won't integrate tightly on all levels (which can be fine).

The other thing is to implement a https://github.com/numpy/numpy-user-dtypes, but yes.  For now that is fully in C.  It isn't particularly hard for this type of DType as long as you have an existing C/C++ library that covers some of the math (of course you can also implement the math, but its just one big chunk less).

**Comment by jorenham** [comment]
Dual numbers are amazing! Here's some more info on them for those interested: https://www.youtube.com/watch?v=ceaNqdHdqtg . They were recently also featured in a computerphile video: https://www.youtube.com/watch?v=QwFLA5TrviI

**Comment by mhvk** [comment]
@ev-br - nice that much of this is already in `scipy` - note to those who might want to pursue this, the `xsf` C code is being made a separate library for the explicit purpose of allowing other packages to use the pieces; see https://github.com/scipy/xsf/issues/1

**Comment by mattip** [comment]
It sounds like SciPy is providing a path to dual numbers. Let's close this and continue discussion there. An alternative may be to add a dual dtype to numpy-user-dtypes.
